pgxn/neon/walproposer.c

```diff
@@ -1601,13 +1618,28 @@ GetDonor(WalProposer *wp, XLogRecPtr *donor_lsn)
 	{
 		Safekeeper *sk = &wp->safekeeper[i];
 
-		if (sk->state == SS_ACTIVE && sk->appendResponse.flushLsn > *donor_lsn)
+		if (sk->state == SS_ACTIVE && sk->appendResponse.flushLsn > donor_lsn)
 		{
 			donor = sk;
-			*donor_lsn = sk->appendResponse.flushLsn;
+			donor_lsn = sk->appendResponse.flushLsn;
 		}
 	}
-	return donor;
+
+        if(donor == NULL)
+        {
+            wp_log(WARNING, "UpdateDonorShmem didn't find a suitable donor, skipping");
+            return;
+        }
+
+        snprintf(donor_name, sizeof(donor_name), "%s:%s", donor->host, donor->port);
+#ifndef WALPROPOSER_LIB
+	SpinLockAcquire(&wps->mutex);
+#endif
+	memcpy(wps->donor_name, donor_name, sizeof(donor_name));
+        memcpy(wps->donor_conninfo, donor->conninfo, sizeof(donor->conninfo));
+#ifndef WALPROPOSER_LIB
+	SpinLockRelease(&wps->mutex);
+#endif
```

Comment:
WDYT about moving this section to `walproposer_pg.c`? I don't like `walproposer.c` accesing shmem directly, my initial idea about separation of those files was that `walproposer.c` can be compiled and used separately as a library, and it doesn't fit well together with calling spinlock here.

My proposal:
- add `void (*update_donor) (WalProposer *wp, Safekeeper *donor, XLogRecPtr donor_lsn);` to `walproposer_api`
- implement it in `walproposer_pg.c`
- call it from here

Reply:
Sure, that works